src/Microsoft.ML.SearchSpace/SearchSpace.cs

```diff
@@ -153,6 +155,32 @@ public override double[] MappingToFeatureSpace(Parameter parameter)
             return res.ToArray();
         }
 
+        public override int GetHashCode()
+        {
+            // hash code is calculated in the following process
+            // 1. sample parameter from search space with all feature value equals to 0.5
+            // 2. serialize sampled parameter to json string
+            // 3. return json string hash code
+
+            var featureSpace = Enumerable.Repeat(0.5, FeatureSpaceDim).ToArray();
+            var parameter = SampleFromFeatureSpace(featureSpace);
+            var json = JsonSerializer.Serialize(parameter);
+
+            // we need to make sure the hash code is the same not only during the same training session, but also
+            // on different platform/CLR, so we can't use string.GetHashCode() here.
+            unchecked
+            {
+                uint hash = 31;
+                foreach (var c in json)
+                {
+                    hash += hash * 7 + c;
+                }
+
+                // make sure hash code is greater than 0
+                return (int)(hash >> 1);
```

Comment:
Is it picked up from Java? :-)

You can simplify that like 
[https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Tuple.cs,69](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Tuple.cs,69)

without using checked blocks at all.

Reply:
That's cool, thanks!

Reply:
### Resolved